### PR TITLE
Refactor/plugins

### PR DESCRIPTION
**Pull request by ndcampbell**
refactored the plugins to use pluginbase and apply a default if not found.

### CODE REVIEW

File: src/artifactorypurge/artifactorypurge.py [hunk 1]
```diff
@@ -1,64 +1,78 @@
...
-            LOG.error("  {}: {}".format(repo, exception))
-
-        exit(1)
-
     LOG.info("Done.")
 
     exit(0)
 
 
+def setup_pluginbase(plugin_path):
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: This should be called `extra_plugin_path` and probably have a default value.
I also think this should be optional.

File: src/artifactorypurge/artifactorypurge.py [hunk 2]
```diff
@@ -1,64 +1,78 @@
...
 LOG = getLogger(__name__)
 
+
 @click.command()
-@click.option('--dryrun/--nodryrun', default=True, is_flag=True, help='Dryrun does not delete any artifacts. On by default')
-@click.option('--reponame', default=None, help='Operate on single repo')
-@click.option('--project', default=None, help='Operate on single project')
+@click.option(
+    '--dryrun/--nodryrun', default=True, is_flag=True, help='Dryrun does not delete any artifacts. On by default')
+@click.option('--plugin-path', required=True, help='Path to plugin directory')
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I believe this should be optional. I should just be able to run it.
A normal person may not care about the plugins. What if they don't have any custom ones? What would they enter?

File: src/artifactorypurge/artifactorypurge.py [hunk 3]
```diff
@@ -1,64 +1,78 @@
...
+        plugin_path (str): Extra path to find plugins in
+
+    Returns:
+        PluginSource: PluginBase PluginSource for finding plugins
+    """
+    here = os.path.dirname(os.path.realpath(__file__))
+    default_path = "{}/plugins".format(here)
+    LOG.info("Searching for plugins in %s and %s", plugin_path, default_path)
+    plugin_base = PluginBase(package='artifactorypurge.plugins')
+    plugin_source = plugin_base.make_plugin_source(searchpath=[plugin_path, default_path])
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: Once `extra_plugin_path` is optional, you'll have to adjust this.
Maybe soemthing like this:
```
+    all_paths = [default_path]
+    if plugin_path:
+        all_paths.append(plugin_path)
```

File: src/artifactorypurge/artifactorypurge.py [hunk 4]
```diff
@@ -1,64 +1,78 @@
...
+    artifactory = Artifactory(credentials['artifactory_url'], credentials['artifactory_username'],
                               credentials['artifactory_password'])
 
-    before = artifactory.list(reponame)
+    plugin_source = setup_pluginbase(plugin_path)
+    before = artifactory.list(None)
     for repo, info in before.items():
-        fp = 0
+        plugin_name = repo.replace("-", "_")
         try:
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I have to look at this tomorrow and give more feedback then.

File: src/artifactorypurge/artifactorypurge.py (hunk 2 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: agreed. I will update that

File: src/artifactorypurge/artifactorypurge.py (hunk 1 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: agreed

**Comment by ndcampbell** [comment]
@sijis I made `policies-path` optional, and I also renamed the arguement to `policies-path` instead of `plugin-path` so that it is more clear what it is searching for. I am going to refer to the plugins as "policies" throughout the code so that their function is clear

File: src/artifactorypurge/artifactorypurge.py [hunk 5]
```diff
@@ -1,64 +1,81 @@
...
-            exceptions[repo] = str(e)
-        except ImportError:
-            continue
-        finally:
-            if fp:
-                fp.close()
+            artifactory_policy = plugin_source.load_plugin(policy_name)
+        except ModuleNotFoundError:
+            if default:
+                LOG.info("No plugin found for %s. Applying Default", repo)
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: Change name to `policy`.

File: src/artifactorypurge/artifactorypurge.py [hunk 6]
```diff
@@ -1,64 +1,81 @@
...
+                artifactory_policy = plugin_source.load_plugin('default')
+            else:
+                LOG.info("No plugin found for %s. Skipping Default", repo)
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: Change to `policy`.

File: src/artifactorypurge/artifactorypurge.py [hunk 7]
```diff
@@ -1,64 +1,81 @@
...
+                continue
+        artifacts = artifactory_policy.purgelist(
+            artifactory,
+            repo,
+            None,
+        )
+        count = artifactory.purge(repo, dryrun, artifacts)
+        LOG.info("processed {}, purged {}".format(repo, count))
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: Use capital P's.

File: src/artifactorypurge/artifactorypurge.py [hunk 8]
```diff
@@ -1,64 +1,81 @@
...
+    Args:
+        extra_policies_path (str): Extra path to find plugins in
+
+    Returns:
+        PluginSource: PluginBase PluginSource for finding plugins
+    """
+    here = os.path.dirname(os.path.realpath(__file__))
+    default_path = "{}/policies".format(here)
+    all_paths = [default_path]
+    if extra_policies_path:
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: We should probably validate that path provided exists too.

File: src/artifactorypurge/artifactorypurge.py [hunk 9]
```diff
@@ -1,64 +1,81 @@
...
-            count = artifactory.purge(repo, dryrun, artifacts)
-            LOG.info("processed {}, purged {}".format(repo, count))
-        except IndexError as e:  # FIXME: return to generic catch
-            exceptions[repo] = str(e)
-        except ImportError:
-            continue
-        finally:
-            if fp:
-                fp.close()
+            artifactory_policy = plugin_source.load_plugin(policy_name)
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I don't think thre's a need to sprinkle 'artifactory' in the variable on everything. Its just a policy.
It's also a little hard to do searches if everything is 'artifactory_'...

File: src/artifactorypurge/artifactorypurge.py [hunk 10]
```diff
@@ -1,64 +1,81 @@
...
+        except ModuleNotFoundError:
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I think the logic is sound.
I would maybe move the `No plugin found for ...` stuff outside the if/else and add a second entry for appying the default policy or not.

File: src/artifactorypurge/artifactorypurge.py [hunk 11]
```diff
@@ -1,64 +1,81 @@
...
-
+def purge(dryrun, policies_path, default):  #, url):
     credentials = load_credentials()
-    artifactory = Artifactory(credentials['artifactory_url'], 
-                              credentials['artifactory_username'],
+    artifactory = Artifactory(credentials['artifactory_url'], credentials['artifactory_username'],
                               credentials['artifactory_password'])
 
-    before = artifactory.list(reponame)
+    plugin_source = setup_pluginbase(extra_policies_path=policies_path)
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I think you'd probably want to add a `LOG.debug` to show you all the plugins found. I'm sure we will need this later when troubleshooting.

File: src/artifactorypurge/artifactorypurge.py [hunk 12]
```diff
@@ -1,64 +1,81 @@
...
+            None,
+        )
+        count = artifactory.purge(repo, dryrun, artifacts)
+        LOG.info("processed {}, purged {}".format(repo, count))
 
     LOG.info("")
     LOG.info("Purging Performance:")
-    after = artifactory.list(reponame)
+    after = artifactory.list(None)
     for repo, info in after.items():
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: Not specific to actual changes, but moving this to a separate function may make sense.

File: src/artifactorypurge/artifactorypurge.py [hunk 13]
```diff
@@ -1,64 +1,81 @@
...
+
+    Returns:
+        PluginSource: PluginBase PluginSource for finding plugins
+    """
+    here = os.path.dirname(os.path.realpath(__file__))
+    default_path = "{}/policies".format(here)
+    all_paths = [default_path]
+    if extra_policies_path:
+        all_paths.append(extra_policies_path)
+    LOG.info("Searching for policies in %s", str(all_paths))
```
Comment (sijis) [inline src/artifactorypurge/artifactorypurge.py]: I'm wondering if `str(all_paths)` should be `', '.join(all_paths)`. it seems to look better locally. More cosmetic'ish.

File: src/artifactorypurge/artifactorypurge.py (hunk 12 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: completely agree, this was Alex code that I just had not refactored yet.

File: src/artifactorypurge/artifactorypurge.py (hunk 13 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: yeah, agreed

File: src/artifactorypurge/artifactorypurge.py (hunk 8 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: Fixed this, added exception

File: src/artifactorypurge/artifactorypurge.py (hunk 11 above)
Comment (ndcampbell) [reply src/artifactorypurge/artifactorypurge.py]: added

**Pull request merged**